internal/controller/service_accounts.go

```diff
@@ -16,14 +17,33 @@ const (
 	modelServerServiceAccountName      = "model-server"
 	notebookServiceAccountName         = "notebook"
 	dataLoaderServiceAccountName       = "data-loader"
+	identityBoundLabel                 = "substratusai/identity-bound"
 )
 
-func reconcileCloudServiceAccount(ctx context.Context, cloudConfig cloud.Cloud, c client.Client, sa *corev1.ServiceAccount) (result, error) {
+func reconcileCloudServiceAccount(ctx context.Context, cloudConfig cloud.Cloud, sciClient sci.ControllerClient, c client.Client, sa *corev1.ServiceAccount) (result, error) {
+	if sa.Annotations == nil {
+		sa.Annotations = map[string]string{}
+	}
+
 	configureSA := func() error {
-		cloudConfig.AssociateServiceAccount(sa)
+		cloudConfig.AssociatePrincipal(sa)
 		return nil
 	}
 
+	principal := cloudConfig.GetPrincipal(sa)
+	if val, exist := sa.Annotations[identityBoundLabel]; !exist || val != principal {
+		bindIdentityRequest := sci.BindIdentityRequest{
+			Principal:                principal,
+			KubernetesServiceAccount: sa.Name,
+			KubernetesNamespace:      sa.Namespace,
+		}
+		if _, err := sciClient.BindIdentity(ctx, &bindIdentityRequest); err != nil {
+			return result{}, fmt.Errorf("failed bind identity principal %s to K8s SA %s/%s: %w",
+				principal, sa.Namespace, sa.Name, err)
+		}
+		sa.Annotations["identity-bound"] = principal
```

Comment:
This is ok, but it would be cleaner to re-use the cloud-specific annotation for the signal that this has been bound. How this could be done without passing around the cloud-specific annotation key? Perhaps `cloudConfig.GetPrinciple(sa)` could return two values `(string, bool)` with the 2nd being: whether this principal was already set on the given ServiceAccount object.

Reply:
I think that would complicate it and make it less clean. It's  clearer what's happening here vs passing around  a label. 
Let me know if you feel strongly about it.

I did have a bug here because I wasn't using the `const` when setting the annotation.

Reply:
I implemented your suggestion and agree that it's cleaner now. I misunderstood your comment originally. Great suggestion!